### PR TITLE
add basic user interface

### PR DESCRIPTION
**Pull request by mcc12**
adds main panel 

adds tab column with fully customise able tabs that may accept images for the tab icons

adds content container for displaying the active tab content

adds column layout which divides active tab content into the components and the optional element label, these are divided into two columns, or one if no label is provided. This way custom components are able to be provided with a label by default

adds row layout which takes the resulting component information and splits it into rows, and the nested columns within those rows. This allows for creation of complex display structures with custom components or the default ones, with many options for customisation.

adds label for displaying information that is not intended to be interactive. Displays as HTML text, so it is able to be copied.

adds text input. This is mutate able into text box, text area, password input and number input. 

adds button input with customise able colours and the option to provide a source image for the button icon

the entries being pushed onto the tab array inside BoltGUI are temporary for testing the interface. These will be removed once we have the ability to generate this array from the information passed into bolt

### CODE REVIEW

File: src/components/BoltGUI.tsx [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        this.state = {
+            expandedX: true,
+            expandedY: true,
+            activeTab: 0,
+        };
+
+        const logInputChange = (value:any) => console.log(value);
+        
+        this.tabs.push({
+            title: 'Example',
```
Comment (maurogb) [inline src/components/BoltGUI.tsx:37]: Just an idea, feel free to ignore if does not make sense: is it possible to put tab data into a separate file, like tab-data js/json? That could provide a good idea on what is the editable part.

File: src/components/ContentPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    panelWidth?: string;
+    /** Information to be rendered in content panel */
+    tabData: TabData;
+}
+
+interface ContainerProps {
+    panelWidth?: string;
+}
+
+const Container = styled.div<ContainerProps>`
```
Comment (maurogb) [inline src/components/ContentPanel.tsx]: Naive question: do we have a plan for theming these styled components? Could we consider these css as default settings that could be overridden by another set of styles?

File: src/components/MainPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+{
+    showScroll: boolean;
+    width?: string;
+}
+
+const Container = styled.div<ContainerProps>`
+    position: fixed;
+    left: 50px;
+    top: 100px;
+    width: ${(props: ContainerProps): string => (props.expandedX ? props.width : '100px')};
```
Comment (maurogb) [inline src/components/MainPanel.tsx]: This looks amazing!

**Review by maurogb — APPROVED** [review]
Really nice one! I've built and ran this one locally because, honestly, I was more learning than reviewing. Things that I've never heard about like the storybook, or seeing all that styled components power in practice... pretty awesome! Also I liked the idea of favouring FunctionComponents over regular components, looks very neat, seems to make easier to develop and test. 
Well... I don't have much more to say... Just looking forward to see the next stages of development! Well done!!

File: src/components/ContentPanel.tsx (hunk 2 above)
Comment (Zyie) [reply src/components/ContentPanel.tsx]: yeah theming is something we will be adding once we get the basic inputs in

File: src/components/BoltGUI.tsx (hunk 1 above)
Comment (mcc12) [reply src/components/BoltGUI.tsx:37]: this is a good idea, we could use it for a default setup, there is a plan to populate the tabs array outside this class in future

**Review by GoodBoyDigital — APPROVED** [review]
this looks really cool as a first pass! 

This already looks like it could be useful and replace Dat.GUI! To that end I would suggest next steps:

- Working toward a first release on npm that we can then create an example of it in use on a test project!
- I would like you to consider how we can make a really clean API for adding elements to the GUI from a project. 

eg:
```
bolt.add(this, 'x', {min:0, max:100, label:'player-position'});

const playerGroup = bolt.addGroup('player options');
playerGroup.addButton(()=>{console.log("pressed"}, {label:'test-button'});
```

Nice work @mcc12 !

**Comment by Zyie** [comment]
> * Working toward a first release on npm that we can then create an example of it in use on a test project!

So i already have a roadmap for what the first release needs for it to be useful. I'll make a github project and we can tick it off as we prepare for the release

> * I would like you to consider how we can make a really clean API for adding elements to the GUI from a project.

This is where i come in. I already have the API somewhat mapped out

**Pull request merged**